File: src/lib/llm/__tests__/thinking-detector.test.ts
/**
 * ThinkingDetector 单元测试
 * 测试 LLM 思考标签检测功能
 */

import { ThinkingDetector } from '../thinking-detector';

describe('ThinkingDetector', () => {
  let detector: ThinkingDetector;

  beforeEach(() => {
    detector = new ThinkingDetector();
  });

  describe('基本功能', () => {
    it('应正确初始化状态', () => {
      const state = detector.getState();
      expect(state.state).toBe('OUTSIDE');
      expect(state.bufferLength).toBe(0);
    });

    it('应正确处理空 chunk', () => {
      const result = detector.process('');
      expect(result.content).toBe('');
      expect(result.reasoning).toBe('');
    });

    it('应正确处理纯文本', () => {
      const result = detector.process('Hello, world!');
      expect(result.content).toBe('Hello, world!');
      expect(result.reasoning).toBe('');
    });

    it('应正确重置状态', () => {
      detector.process('<think>thinking</think>');
      detector.reset();
      const state = detector.getState();
      expect(state.state).toBe('OUTSIDE');
      expect(state.bufferLength).toBe(0);
    });
  });

  describe('<think> 标签检测', () => {
    it('应正确检测标准 <think> 标签', () => {
      const result = detector.process('<think>This is thinking</think>');
      expect(result.content).toBe('');
      expect(result.reasoning).toBe('This is thinking');
    });

    it('应正确检测带空格的 <think> 标签', () => {
      const result = detector.process('<think >Thinking content</think>');
      expect(result.reasoning).toBe('Thinking content');
    });

    it('应正确检测 </think> 结束标签', () => {
      const result = detector.process('<think>inner</think>more text');
      expect(result.reasoning).toBe('inner');
      expect(result.content).toBe('more text');
    });

    it('应正确处理分段到达的 <think> 标签', () => {
      detector.process('<th');
      const result = detector.process('ink>thinking</think>');
      expect(result.reasoning).toBe('thinking');
    });

    it('应正确处理带属性的 <think> 标签', () => {
      const result = detector.process('<think type="reasoning">content</think>');
      expect(result.reasoning).toBe('content');
    });

    it('应正确处理嵌套标签', () => {
      const result = detector.process('<think>outer <think>inner</think> back</think>');
      expect(result.reasoning).toContain('outer');
      expect(result.reasoning).toContain('inner');
    });
  });

  describe('<thought> 标签检测', () => {
    it('应正确检测 <thought> 标签', () => {
      const result = detector.process('<thought>Thoughts here</thought>');
      expect(result.reasoning).toBe('Thoughts here');
    });

    it('应正确检测带空格的 <thought> 标签', () => {
      const result = detector.process('<thought >Content</thought>');
      expect(result.reasoning).toBe('Content');
    });

    it('应正确检测 </thought> 结束标签', () => {
      const result = detector.process('<thought>reasoning</thought>normal text');
      expect(result.reasoning).toBe('reasoning');
      expect(result.content).toBe('normal text');
    });
  });

  describe('HTML 注释格式检测', () => {
    it('应正确检测 <!-- THINKING_START -->', () => {
      const result = detector.process('<!-- THINKING_START -->thinking<!-- THINKING_END -->');
      expect(result.reasoning).toBe('thinking');
    });

    it('应正确检测带空格的变体', () => {
      const result = detector.process('<!-- THINKING_START -->thoughts<!-- THINKING_END -->');
      expect(result.reasoning).toBe('thoughts');
    });

    it('应正确处理多行思考内容', () => {
      const result = detector.process(
        '<!-- THINKING_START -->Line 1\nLine 2\nLine 3<!-- THINKING_END -->'
      );
      expect(result.reasoning).toContain('Line 1');
      expect(result.reasoning).toContain('Line 2');
      expect(result.reasoning).toContain('Line 3');
    });

    it('应正确处理混合格式', () => {
      const result = detector.process(
        'Before <!-- THINKING_START -->thought<!-- THINKING_END --> After'
      );
      expect(result.content).toContain('Before');
      expect(result.content).toContain('After');
      expect(result.reasoning).toBe('thought');
    });
  });

  describe('混合格式处理', () => {
    it('应正确处理文本+思考+文本', () => {
      const result = detector.process('Hello <think>reasoning</think> World');
      expect(result.content).toContain('Hello');
      expect(result.content).toContain('World');
      expect(result.reasoning).toBe('reasoning');
    });

    it('应正确处理多段思考', () => {
      const result = detector.process(
        'Start <think>T1</think> middle <think>T2</think> end'
      );
      expect(result.reasoning).toContain('T1');
      expect(result.reasoning).toContain('T2');
      expect(result.content).toContain('Start');
      expect(result.content).toContain('middle');
      expect(result.content).toContain('end');
    });

    it('应正确处理工具调用+思考+工具调用', () => {
      const result = detector.process(
        '<tool_code>{}</tool_code><think>thinking</think><tool_code>{}</tool_code>'
      );
      // ThinkingDetector 只处理思考标签，不过滤工具调用
      expect(result.reasoning).toBe('thinking');
    });
  });

  describe('分段处理', () => {
    it('应正确处理分段到达的内容', () => {
      const r1 = detector.process('<think>part');
      const result = detector.process('ial thinking</think>');
      const allReasoning = r1.reasoning + result.reasoning;
      expect(allReasoning).toContain('partial');
    });

    it('应正确处理分段结束标签', () => {
      const r1 = detector.process('<think>thoughts');
      const result = detector.process('</think>more');
      const allReasoning = r1.reasoning + result.reasoning;
      expect(allReasoning).toContain('thoughts');
      expect(result.content).toContain('more');
    });

    it('应正确处理跨多段的思考', () => {
      const r1 = detector.process('<think>1');
      const r2 = detector.process('2');
      const r3 = detector.process('3');
      const r4 = detector.process('</think>done');
      const allReasoning = r1.reasoning + r2.reasoning + r3.reasoning + r4.reasoning;
      expect(allReasoning).toContain('123');
      expect(r4.content).toContain('done');
    });

    it('应正确处理分段到达的 HTML 注释格式', () => {
      detector.process('<!-- THINKING_');
      detector.process('START -->');
      const result = detector.process('thinking<!-- THINKING_END -->');
      expect(result.reasoning).toContain('thinking');
    });
  });

  describe('flush 功能', () => {
    it('应正确 flush 未关闭的思考块', () => {
      const r1 = detector.process('<think>unclosed thinking');
      const result = detector.flush();
      const allReasoning = r1.reasoning + result.reasoning;
      expect(allReasoning).toContain('unclosed');
      expect(typeof result.content).toBe('string');
    });

    it('flush 后状态应重置', () => {
      detector.process('<think>thinking');
      detector.flush();
      const state = detector.getState();
      expect(state.state).toBe('OUTSIDE');
    });

    it('flush 应清空缓冲区', () => {
      detector.process('some text');
      detector.flush();
      const state = detector.getState();
      expect(state.bufferLength).toBe(0);
    });

    it('flush 应处理正常内容', () => {
      detector.process('normal text');
      const result = detector.flush();
      // process() 已将内容输出，flush() 输出剩余 buffer（可能为空）
      expect(typeof result.content).toBe('string');
    });
  });

  describe('边界情况', () => {
    it('应处理连续的开始标签', () => {
      const result = detector.process('<think>T1</think><think>T2</think>');
      expect(result.reasoning).toContain('T1');
      expect(result.reasoning).toContain('T2');
    });

    it('应处理空的思考块', () => {
      const result = detector.process('<think></think>');
      expect(result.reasoning).toBe('');
    });

    it('应处理只有空白符的思考块', () => {
      const result = detector.process('<think>   \n  </think>');
      expect(result.reasoning.trim()).toBe('');
    });

    it('应处理超长思考内容', () => {
      const longThinking = 'x'.repeat(10000);
      const result = detector.process(`<think>${longThinking}</think>`);
      expect(result.reasoning).toBe(longThinking);
    });

    it('应处理 Unicode 内容', () => {
      const result = detector.process('<think>中文思考🤔</think>');
      expect(result.reasoning).toBe('中文思考🤔');
    });

    it('应处理特殊字符', () => {
      const result = detector.process('<think><>&"\'</think>');
      expect(result.reasoning).toContain('<');
      expect(result.reasoning).toContain('>')
    });
  });

  describe('状态追踪', () => {
    it('应正确追踪 OUTSIDE 状态', () => {
      detector.process('normal text');
      const state = detector.getState();
      expect(state.state).toBe('OUTSIDE');
    });

    it('应正确追踪 INSIDE 状态', () => {
      detector.process('<think>in progress');
      const state = detector.getState();
      expect(state.state).toBe('INSIDE');
    });

    it('应正确追踪缓冲区长度', () => {
      detector.process('hello');
      const state = detector.getState();
      expect(state.bufferLength).toBeGreaterThanOrEqual(0);
    });
  });

  describe('缓冲区安全处理', () => {
    it('应保留尾部 < 字符', () => {
      detector.process('text <');
      const state = detector.getState();
      expect(state.bufferLength).toBeGreaterThan(0);
    });

    it('应保留尾部 <! 字符', () => {
      detector.process('text <!');
      const state = detector.getState();
      expect(state.bufferLength).toBeGreaterThan(0);
    });

    it('应保留尾部 <!-- 字符', () => {
      detector.process('text <!--');
      const state = detector.getState();
      expect(state.bufferLength).toBeGreaterThan(0);
    });

    it('不应无限循环', () => {
      const start = Date.now();
      // 传入大量文本应能在有限时间内完成
      detector.process('a'.repeat(10000));
      const elapsed = Date.now() - start;
      expect(elapsed).toBeLessThan(1000);
    });
  });
});

File: native-ui/app/src/test/java/com/promenar/nexara/data/remote/ThinkingDetectorTest.kt
package com.promenar.nexara.data.remote

import com.google.common.truth.Truth.assertThat
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test

@DisplayName("ThinkingDetector")
class ThinkingDetectorTest {

    private lateinit var detector: ThinkingDetector

    @BeforeEach
    fun setUp() {
        detector = ThinkingDetector()
    }

    @Nested
    @DisplayName("Plain text (no tags)")
    inner class PlainText {

        @Test
        @DisplayName("single chunk passes through as content")
        fun singleChunk() {
            val result = detector.process("Hello, world!")
            assertThat(result.content).isEqualTo("Hello, world!")
            assertThat(result.reasoning).isEmpty()
        }

        @Test
        @DisplayName("multiple chunks accumulate as content")
        fun multipleChunks() {
            val r1 = detector.process("Hello")
            val r2 = detector.process(", ")
            val r3 = detector.process("world!")
            assertThat(r1.content + r2.content + r3.content).isEqualTo("Hello, world!")
            assertThat(r1.reasoning + r2.reasoning + r3.reasoning).isEmpty()
        }

        @Test
        @DisplayName("empty chunk returns empty result")
        fun emptyChunk() {
            val result = detector.process("")
            assertThat(result.content).isEmpty()
            assertThat(result.reasoning).isEmpty()
        }
    }

    @Nested
    @DisplayName("<think/> tags")
    inner class ThinkTags {

        @Test
        @DisplayName("single think block in one chunk")
        fun singleBlock() {
            val result = detector.process("before<think/>reasoning here</think/>after")
            assertThat(result.content).isEqualTo("beforeafter")
            assertThat(result.reasoning).isEqualTo("reasoning here")
        }

        @Test
        @DisplayName("think block with attributes")
        fun withAttributes() {
            val result = detector.process("before<think type=\"reasoning\">inner thoughts</think/>after")
            assertThat(result.content).isEqualTo("beforeafter")
            assertThat(result.reasoning).isEqualTo("inner thoughts")
        }

        @Test
        @DisplayName("multiple think blocks")
        fun multipleBlocks() {
            val r1 = detector.process("a<think/>b</think/>c<think/>d</think/>e")
            assertThat(r1.content).isEqualTo("ace")
            assertThat(r1.reasoning).isEqualTo("bd")
        }
    }

    @Nested
    @DisplayName("<thought/> tags")
    inner class ThoughtTags {

        @Test
        @DisplayName("single thought block")
        fun singleBlock() {
            val result = detector.process("before<thought/>deep thinking</thought/>after")
            assertThat(result.content).isEqualTo("beforeafter")
            assertThat(result.reasoning).isEqualTo("deep thinking")
        }

        @Test
        @DisplayName("thought block with attributes")
        fun withAttributes() {
            val result = detector.process("a<thought level=2>reasoning</thought>b")
            assertThat(result.content).isEqualTo("ab")
            assertThat(result.reasoning).isEqualTo("reasoning")
        }
    }

    @Nested
    @DisplayName("<!-- THINKING_START/END --> comments")
    inner class HtmlCommentTags {

        @Test
        @DisplayName("comment-style thinking tags")
        fun commentStyle() {
            val result = detector.process("before<!-- THINKING_START -->thinking<!-- THINKING_END -->after")
            assertThat(result.content).isEqualTo("beforeafter")
            assertThat(result.reasoning).isEqualTo("thinking")
        }

        @Test
        @DisplayName("comment tags with extra whitespace")
        fun commentWithWhitespace() {
            val result = detector.process("a<!--  THINKING_START  -->thoughts<!--  THINKING_END  -->b")
            assertThat(result.content).isEqualTo("ab")
            assertThat(result.reasoning).isEqualTo("thoughts")
        }

        @Test
        @DisplayName("comment tags case insensitive")
        fun commentCaseInsensitive() {
            val result = detector.process("a<!-- thinking_start -->thoughts<!-- thinking_end -->b")
            assertThat(result.content).isEqualTo("ab")
            assertThat(result.reasoning).isEqualTo("thoughts")
        }
    }

    @Nested
    @DisplayName("Cross-chunk boundary handling")
    inner class CrossChunkBoundary {

        @Test
        @DisplayName("open tag split across two chunks")
        fun openTagSplit() {
            val r1 = detector.process("hello<thin")
            val r2 = detector.process("k/>reasoning</think/>world")
            assertThat(r1.content).isEqualTo("hello")
            assertThat(r1.reasoning).isEmpty()
            assertThat(r2.content).isEqualTo("world")
            assertThat(r2.reasoning).isEqualTo("reasoning")
        }

        @Test
        @DisplayName("close tag split across two chunks")
        fun closeTagSplit() {
            val r1 = detector.process("hello<think/>reasoning</thi")
            val r2 = detector.process("nk/>world")
            assertThat(r1.content).isEqualTo("hello")
            assertThat(r1.reasoning).isEqualTo("reasoning")
            assertThat(r2.content).isEqualTo("world")
            assertThat(r2.reasoning).isEmpty()
        }

        @Test
        @DisplayName("entire think block delivered char by char")
        fun charByChar() {
            val text = "a<think/>b</think/>c"
            var allContent = ""
            var allReasoning = ""
            for (ch in text) {
                val r = detector.process(ch.toString())
                allContent += r.content
                allReasoning += r.reasoning
            }
            val final = detector.flush()
            allContent += final.content
            allReasoning += final.reasoning
            assertThat(allContent).isEqualTo("ac")
            assertThat(allReasoning).isEqualTo("b")
        }

        @Test
        @DisplayName("comment open tag split across chunks")
        fun commentOpenTagSplit() {
            val r1 = detector.process("hello<!-- THI")
            val r2 = detector.process("NKING_START -->thoughts<!-- THINKING_END -->world")
            assertThat(r1.content).isEqualTo("hello")
            assertThat(r1.reasoning).isEmpty()
            assertThat(r2.content).isEqualTo("world")
            assertThat(r2.reasoning).isEqualTo("thoughts")
        }

        @Test
        @DisplayName("reasoning content split across many chunks inside think block")
        fun reasoningAcrossChunks() {
            val r1 = detector.process("start<think/>")
            val r2 = detector.process("part1")
            val r3 = detector.process("part2")
            val r4 = detector.process("</think/>end")
            assertThat(r1.content).isEqualTo("start")
            assertThat(r1.reasoning).isEmpty()
            assertThat(r2.content).isEmpty()
            assertThat(r2.reasoning).isEqualTo("part1")
            assertThat(r3.content).isEmpty()
            assertThat(r3.reasoning).isEqualTo("part2")
            assertThat(r4.content).isEqualTo("end")
            assertThat(r4.reasoning).isEmpty()
        }

        @Test
        @DisplayName("bracket at end of chunk preserved in buffer")
        fun bracketAtEndPreserved() {
            val r1 = detector.process("some text<")
            assertThat(r1.content).isEqualTo("some text")
            assertThat(r1.reasoning).isEmpty()
            val r2 = detector.process("more text")
            assertThat(r2.content).isEmpty()
            assertThat(r2.reasoning).isEmpty()
            val r3 = detector.flush()
            assertThat(r3.content).isEqualTo("<more text")
            assertThat(r3.reasoning).isEmpty()
        }
    }

    @Nested
    @DisplayName("Flush behavior")
    inner class Flush {

        @Test
        @DisplayName("flush outputs remaining buffer as content when OUTSIDE")
        fun flushOutside() {
            detector.process("some content<")
            val result = detector.flush()
            assertThat(result.content).isEqualTo("<")
            assertThat(result.reasoning).isEmpty()
        }

        @Test
        @DisplayName("flush outputs remaining buffer as reasoning when INSIDE")
        fun flushInside() {
            val r1 = detector.process("before<think/>unfinished reasoning<")
            assertThat(r1.content).isEqualTo("before")
            assertThat(r1.reasoning).isEqualTo("unfinished reasoning")
            val result = detector.flush()
            assertThat(result.content).isEmpty()
            assertThat(result.reasoning).isEqualTo("<")
        }

        @Test
        @DisplayName("flush resets state")
        fun flushResets() {
            detector.process("before<think/>reasoning")
            detector.flush()
            val state = detector.getState()
            assertThat(state.state).isEqualTo("OUTSIDE")
            assertThat(state.bufferLength).isEqualTo(0)
        }

        @Test
        @DisplayName("flush after flush returns empty")
        fun doubleFlush() {
            detector.process("content")
            detector.flush()
            val result = detector.flush()
            assertThat(result.content).isEmpty()
            assertThat(result.reasoning).isEmpty()
        }
    }

    @Nested
    @DisplayName("Reset behavior")
    inner class Reset {

        @Test
        @DisplayName("reset clears all state")
        fun resetClears() {
            detector.process("before<think/>reasoning")
            detector.reset()
            val state = detector.getState()
            assertThat(state.state).isEqualTo("OUTSIDE")
            assertThat(state.bufferLength).isEqualTo(0)
        }

        @Test
        @DisplayName("process after reset starts fresh")
        fun processAfterReset() {
            detector.process("before<think/>reasoning")
            detector.reset()
            val result = detector.process("fresh content")
            assertThat(result.content).isEqualTo("fresh content")
            assertThat(result.reasoning).isEmpty()
        }
    }

    @Nested
    @DisplayName("Edge cases")
    inner class EdgeCases {

        @Test
        @DisplayName("empty think block")
        fun emptyThinkBlock() {
            val result = detector.process("before<think/></think/>after")
            assertThat(result.content).isEqualTo("beforeafter")
            assertThat(result.reasoning).isEmpty()
        }

        @Test
        @DisplayName("nested angle brackets in reasoning")
        fun angleBracketsInReasoning() {
            val result = detector.process("before<think/>a < b && c > d</think/>after")
            assertThat(result.content).isEqualTo("beforeafter")
            assertThat(result.reasoning).isEqualTo("a < b && c > d")
        }

        @Test
        @DisplayName("think tag with whitespace after name")
        fun thinkTagWithWhitespace() {
            val result = detector.process("before<think >reasoning</think >after")
            assertThat(result.content).isEqualTo("beforeafter")
            assertThat(result.reasoning).isEqualTo("reasoning")
        }

        @Test
        @DisplayName("close tag with extra whitespace")
        fun closeTagWithWhitespace() {
            val result = detector.process("before<think/>reasoning</think  >after")
            assertThat(result.content).isEqualTo("beforeafter")
            assertThat(result.reasoning).isEqualTo("reasoning")
        }

        @Test
        @DisplayName("mixed tag formats in one stream")
        fun mixedTagFormats() {
            val result = detector.process(
                "a<think/>b</think/>c<!-- THINKING_START -->d<!-- THINKING_END -->e<thought/>f</thought/>g"
            )
            assertThat(result.content).isEqualTo("aceg")
            assertThat(result.reasoning).isEqualTo("bdf")
        }

        @Test
        @DisplayName("very long reasoning content")
        fun longReasoning() {
            val reasoning = "x".repeat(10000)
            val result = detector.process("before<think/>${reasoning}</think/>after")
            assertThat(result.content).isEqualTo("beforeafter")
            assertThat(result.reasoning).isEqualTo(reasoning)
        }

        @Test
        @DisplayName("unclosed think block - process + flush combine correctly")
        fun unclosedThinkFlushedAsReasoning() {
            val r1 = detector.process("before<think/>partial reasoning")
            assertThat(r1.content).isEqualTo("before")
            assertThat(r1.reasoning).isEqualTo("partial reasoning")
            val flushed = detector.flush()
            assertThat(flushed.content).isEmpty()
            assertThat(flushed.reasoning).isEmpty()
        }

        @Test
        @DisplayName("only think block with no surrounding content")
        fun onlyThinkBlock() {
            val result = detector.process("<think/>just reasoning</think/>")
            assertThat(result.content).isEmpty()
            assertThat(result.reasoning).isEqualTo("just reasoning")
        }
    }

    @Nested
    @DisplayName("Channel-based processFlow")
    inner class ChannelProcessFlow {

        @Test
        @DisplayName("processes chunks via channels with backpressure")
        fun channelProcessing() = runBlocking {
            val input = Channel<String>(10)
            val output = Channel<ThinkingResult>(10)

            launch {
                detector.processFlow(input, output)
            }

            input.send("Hello ")
            input.send("<think/>reasoning")
            input.send("</think/>World")
            input.close()

            val results = mutableListOf<ThinkingResult>()
            withTimeout(2000) {
                for (r in output) {
                    results.add(r)
                }
            }

            val allContent = results.joinToString("") { it.content }
            val allReasoning = results.joinToString("") { it.reasoning }
            assertThat(allContent).isEqualTo("Hello World")
            assertThat(allReasoning).isEqualTo("reasoning")
        }

        @Test
        @DisplayName("channel flushes remaining content on close")
        fun channelFlushOnClose() = runBlocking {
            val input = Channel<String>(10)
            val output = Channel<ThinkingResult>(10)

            launch {
                detector.processFlow(input, output)
            }

            input.send("just content")
            input.close()

            val results = mutableListOf<ThinkingResult>()
            withTimeout(2000) {
                for (r in output) {
                    results.add(r)
                }
            }

            val allContent = results.joinToString("") { it.content }
            assertThat(allContent).isEqualTo("just content")
        }

        @Test
        @DisplayName("channel handles empty input gracefully")
        fun channelEmptyInput() = runBlocking {
            val input = Channel<String>(10)
            val output = Channel<ThinkingResult>(10)

            launch {
                detector.processFlow(input, output)
            }

            input.close()

            val results = mutableListOf<ThinkingResult>()
            withTimeout(2000) {
                for (r in output) {
                    results.add(r)
                }
            }

            assertThat(results).isEmpty()
        }
    }
}
